File: task-test-B.js
const a = readline().split('\r')
  .slice(1)
  .reduce((b, c) => c.toLowerCase()
    .split(' reposted ')
    .reduce((e, g) => {
      if (g !== 'polycarp') {
        if (e[g]) {
          e[g] += 1;
        } else {
          e[g] = 1;
        }
      }

      return e;
    }, b)
  , {});

print(Object.keys(a).reduce((h, i) => a[i] > 1 ? h + 1 : h, 2).toString());

File: task-test-B.py
from collections import defaultdict

def DFS(G,v,seen=None,path=None):
    if seen is None: seen = []
    if path is None: path = [v]

    seen.append(v)

    paths = []
    for t in G[v]:
        if t not in seen:
            t_path = path + [t]
            paths.append(tuple(t_path))
            paths.extend(DFS(G, t, seen[:], t_path))
    return paths

edges = []

for i in xrange(int(raw_input())):
    edges.append(map(lambda x: x.lower(), raw_input().split(' reposted ')))

G = defaultdict(list)

for (s,t) in edges:
    G[s].append(t)
    G[t].append(s)

all_paths = DFS(G, 'polycarp')
max_len   = max(len(p) for p in all_paths)

print(max_len)
